server/sandbox_run.go

```diff
@@ -526,15 +526,19 @@ func (s *Server) RunPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
 
 	var ip string
 	if s.config.Config.ManageNetworkNSLifecycle {
-		ip, err = s.networkStart(sb)
-		if err != nil {
-			return nil, err
-		}
+		// Ensure network resources are cleaned up even if the plugin
+		// succeeded but an error happened between that success and
+		// here.
 		defer func() {
 			if err != nil {
 				s.networkStop(sb)
```

Comment:
Why set this up *before* calling `networkStart`?  I'd have expected:

```go
ip, err = s.networkStart(sb)
if err != nil {
  return nil, err
}
defer func() {
  if err != nil {
    s.networkStop(sb)
  }
}
```

The only reason I can think of for early `defer` would be if `networkStart` explicitly didn't clean up after itself on failure, and since `networkStart` is our code, I'd rather fix that (if it's happening) by patching `networkStart` instead of using an early `defer` here.

Also, it would be nice to log `networkStop` errors instead of silently ignoring them.

Reply:
@wking that's the point of this PR, as explained in the commit message and PR message.

networkStart() does not clean up after itself, nor should it.  I don't think we want networkStart() to be in the business of calling networkStop().  I explicitly left that to the thing that calls networkStart().

Reply:
> @wking that's the point of this PR, as explained in the commit message and PR message.

Ah, my understanding from those messages was that we were trying to clean up from “`networkStart` succeeded but something later in `RunPodSandbox` died”.

> `networkStart()` does not clean up after itself, nor should it. I don't think we want `networkStart()` to be in the business of calling `networkStop()`.

That would be the coarse fix.  A more fine-grained approach in `networkStart` would be:

```go
podNetwork := newPodNetwork(sb)
err := s.netPlugin.SetUpPod(podNetwork)
if err != nil {
  return "", fmt.Errorf("failed to create pod network sandbox %s(%s): %v", sb.Name(), sb.ID(), err)
}
defer func() {
  if err != nil {
    err2 := s.netPlugin.TearDownPod(podNetwork); err != nil {
      logrus.Warnf("failed to destroy network for pod sandbox %s(%s): %v",
        sb.Name(), sb.ID(), err2)
  }
}()
```

Assuming `hostportManager.Add` also cleans up after itself on failures, we wouldn't need a `networkStart` cleanup for that.  Can you explain why you prefer to avoid the deferred `TearDownPod` in `networkStart`?

Reply:
> Also, it would be nice to log networkStop errors instead of silently ignoring them.

networkStop() actually only returns nil but logs errors internally, I'll remove the error return from it.

Reply:
@wking anyway, reworked to put the cleanup logic into networkStart() instead...